реализация модальных окон + получение ингредиентов по api

a1exTee opened the pull request.

tihonovalexey29-star committed:
реализация модальных окон + получение ингредиентов по api

andryxxa93 commented on src/App.tsx:
```diff
@@ -24,8 +24,8 @@ function App() {
     fetch(apiUrl)
-        .then(res => res.json())
-        .then(data => setState({ data: data, isLoading: false, hasError: false }))
+        .then(res => res.json() as Promise<{sucsess: Boolean; data: []}>)
```
**Надо исправить**
Добавьте проверку ответа сервера - res.ok:

Например:

```
.then(res => res.ok ? res.json() : res.json().then((err) => Promise.reject(err));
```

иначе сейчас, если сервер ответи не кодом 2xx, а какой либо ошибкой (4xx, 5xx) мы не попадем в catch т.к. в этом случае fetch не выбрасывает ошибку, мы попытаемся отрисовать данные, хотя они не пришли, в итоге на попытке отрисовки несуществующих данных упадет все приложение

[https://learn.javascript.ru/fetch](https://learn.javascript.ru/fetch)

[https://developer.mozilla.org/en-US/docs/Web/API/Response/ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok)

Т.к. эта проверка нужна у всех запросов, лучше вынести её в отдельную функцию

```
const checkReponse = (res) => {
  return res.ok ? res.json() : res.json().then((err) => Promise.reject(err));
};
```

`fetch(someUrl).then(checkReponse)`

andryxxa93 commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -25,21 +30,65 @@ const BurgerIngredients = ({ingredientsData}) => {
+          <div>
+              <h2>Булки</h2>
+              <div>
+                  {ingredientsData.map((ingredient, index) => {
```
**Можно лучше**

Лучше в компоненте делать filter по массиву, а потом отображать полученный массив.

Например:

```
const buns = useMemo(data.filter((item) => item.type === 'bun'), [data]);
const mains = useMemo(data.filter((item) => item.type === 'main'), [data]);
const sauces = useMemo(data.filter((item) => item.type === 'sauce'), [data]);

return (
   ....................
   {buns.map(....................)}
   ..................
)
```

А не располагать фильтрацию в самом jsx

andryxxa93 commented on src/components/burger-ingredients/ingredients-details/ingredients-details.jsx:
```diff
@@ -2,24 +2,24 @@ import React from 'react';
-const IngredientsDetails = (props: { image: string | undefined; name: string | undefined; proteins: number | undefined; fat: number | undefined; carbohydrates: number | undefined; calories: number | undefined; }) => {
-
+const IngredientsDetails = ({ingredientData}) => {
+console.log(ingredientData);
```
**Можно улучшить**
Console.log нужен только для процесса разработки, перед сдачей проекта его нужно удалять, иначе она будет загружать консоль ненужной, а возможно и приватной информацией. А ошибку лучше записывать в state и показывать пользователю, чтобы он понимал, что происходит.

andryxxa93 commented on src/components/burger-ingredients/ingredients-item/ingredients-item.jsx:
```diff
@@ -2,35 +2,38 @@ import React, { useState } from 'react';
+}; */
+
+IngredientsItem.propTypes = {
+    item: PropTypes.object.isRequired,
```
**Необходимо исправить**
Сейчас получается, что под `item ` может прийти объект любого содержания, но это не так. Необходимо полностью расписать тип ингредиента со всеми его полями (чтобы не дублировать код - сделайте это в отдельном файле) и использовать во всех подобных местах. Вам помогут в этом функции `propTypes` `arrayOf` и `shape`.

andryxxa93 reviewed (changes requested):
Здравствуйте! Проделана большая работа 👍 

Что понравилось:
- Добавили `ErrorBoundary`
- Хорошая декомпозиция компонентов
- Записываете ошибку в стейт

Однако есть пару моментов, которые необходимо исправить, чтобы работа была принята.

Необходимо исправить:

- нужно убрать синюю обводку у блоков, эта обводка на макете нужна только что бы сделать поясняющее описание к этому блоку
- Добавьте проверку ответа сервера - res.ok
- Верхние и нижние булки не должны скролиться в конструкторе.
- Модальные окна не соответствуют макету, посмотрите как в фигме пожалуйста

<img width="544" alt="Снимок экрана 2023-06-04 в 06 44 38" src="https://github.com/a1exTee/react-burger/assets/60173362/9ae19d8f-b122-40a7-a835-4aa474fc673e">
<img width="532" alt="Снимок экрана 2023-06-04 в 06 44 26" src="https://github.com/a1exTee/react-burger/assets/60173362/c6d524ef-5c1e-4f31-a244-34e7b174cb8e">

- В `BurgerIngredients ` и других компонентах неправильно указаны `propTypes`. Пожалуйста проверьте во всём проекте.
- Распишите тип ингредиента и используйте везде в `propTypes`
- Файл App.test.tsx необходимо удалить - он не понадобится.
- У вас сочетаются две технологии типизации - `typescript` и `propTypes` - необходимо оставить одну. Если это `propTypes`, то все файлы с расширением `tsx`  необходимо сделать `jsx`.
- В App и других компонентах вы используете атрибут `style` у тегов - его необходимо использовать для динамических параметров стилей, а во всех остальных случаях использовать классы.

Более подробно описание некотрых проблемы смотрите комментарии в pull request.

В pull request вы так же найдете комментарии с рекомендациями как **можно улучшить** код. Их делать не обязательно, однако в будущем они могут помочь избежать ошибок и сделать код лучше.

**Напомню, чтобы работа была принята нужно исправить все замечания помеченные как **необходимо исправить**.

Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка F12) перед отправкой на ревью. 

Удачных правок!

tihonovalexey29-star committed:
исправлено: -нужно убрать синюю обводку у блоков, эта обводка на макете нужна только что бы сделать поясняющее описание к этому блоку;
-Добавьте проверку ответа сервера - res.ok;
-Верхние и нижние булки не должны скролиться в конструкторе;
-Файл App.test.tsx необходимо удалить - он не понадобится;
-У вас сочетаются две технологии типизации - typescript и propTypes. Необходимо оставить одну. Если это propTypes, то все файлы с расширением tsx необходимо сделать jsx;
-В App и других компонентах вы используете атрибут style у тегов - его необходимо использовать для динамических параметров стилей, а во всех остальных случаях использовать классы;

tihonovalexey29-star committed:
исправлено: Модальные окна не соответствуют макету, посмотрите как в фигме пожалуйста
https://github.com/a1exTee/react-burger/assets/60173362/9ae19d8f-b122-40a7-a835-4aa474fc673e
https://github.com/a1exTee/react-burger/assets/60173362/c6d524ef-5c1e-4f31-a244-34e7b174cb8e
В BurgerIngredients  и других компонентах неправильно указаны propTypes. Пожалуйста проверьте во всём проекте.
Распишите тип ингредиента и используйте везде в propTypes

andryxxa93 commented on src/components/app/App.jsx:
```diff
@@ -0,0 +1,125 @@
+  );
+}
+
+/*function App() {
```
**Можно улучшить**
Ненужный код лучше удалять

andryxxa93 commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -25,21 +30,85 @@ const BurgerIngredients = ({ingredientsData}) => {
 
+
+BurgerIngredients: PropTypes.shape({
+        image: PropTypes.string.isRequired,
```
**Надо исправить**

во многих компонентах дублируется описание PropTypes ингредиента
Следует вынести его в отдельный файл (можно создать файл prop-types.js в папке utils) и переиспользовать описание пропсов ингрединта, а не дублировать

andryxxa93 reviewed (changes requested):
**Необходимо исправить**
- Распишите тип ингредиента и используйте везде в propTypes вместо `propTypes.object` 
- Синяя обводка осталась у блоков с ингредиентами в списке
- Категории в списке ингредиентов не должны скролиться, скролиться должен только сам список ингредиентов

<img width="914" alt="Снимок экрана 2023-06-05 в 09 09 22" src="https://github.com/a1exTee/react-burger/assets/60173362/24fcecd8-24ee-4ffe-b062-576b54d6ab05">
- В проекте много закоментированого кода - его следует удалить
- во многих компонентах дублируется описание PropTypes ингредиента
Следует вынести его в отдельный файл (можно создать файл prop-types.js в папке utils) и переиспользовать описание пропсов ингрединта, а не дублировать

tihonovalexey29-star committed:
- Распишите тип ингредиента и используйте везде в propTypes вместо propTypes.object
- Синяя обводка осталась у блоков с ингредиентами в списке
- Категории в списке ингредиентов не должны скролиться, скролиться должен только сам список ингредиентов
https://github.com/a1exTee/react-burger/assets/60173362/24fcecd8-24ee-4ffe-b062-576b54d6ab05
- В проекте много закоментированого кода - его следует удалить
- во многих компонентах дублируется описание PropTypes ингредиента
Следует вынести его в отдельный файл (можно создать файл prop-types.js в папке utils) и переиспользовать описание пропсов ингрединта, а не дублировать

andryxxa93 commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -70,45 +68,15 @@ const BurgerIngredients = ({ingredientsData}) => {
-    }).isRequired).isRequired,
-};*/
+BurgerIngredients.propTypes = {
+    data: PropTypes.arrayOf(PropTypes.shape(dataPropTypes).isRequired),
```
**Можно лучше**
После arrayOf необходимо так же ставить `isRequired`

andryxxa93 reviewed (approved):
Супер, вы отлично поработали!
Поздравляю! Работа принята.
Желаю удачного продолжения обучения!
